README.md
<a id="devex-badge" rel="Inspiration" href="https://github.com/BCDevExchange/assets/blob/master/README.md"><img alt="An idea being explored and shaped. Open for discussion, but may never go anywhere." style="border-width:0" src="https://assets.bcdevexchange.org/images/badges/inspiration.svg" title="An idea being explored and shaped. Open for discussion, but may never go anywhere." /></a>

[![Travis build
status](https://travis-ci.org/bcgov/geometr.svg?branch=master)](https://travis-ci.org/bcgov/geometr)

# geometr

An R package to query Environment and Climate Change Canada’s GeoMet
webservices

## Project Status

This package is maintained by the Knowledge Management Branch of the
[British Columbia Ministry of Environment and Climate Change
Strategy](https://www2.gov.bc.ca/gov/content/governments/organizational-structure/ministries-organizations/ministries/environment-climate-change).
This project is under construction is not stable for any kind of
sustained use.

### Features

  - Currently the hydrometric historical webservice is the only data
    available through this service

### Installation

``` r
install.packages("remotes")
remotes::install_github("bcgov/geometr")
```

### Usage

First we have to load the package

``` r
library(geometr)
```

The collections current supported by the geomet webservice are (as of
2018-09-18):

``` r
collections <- geomet_collection()

collections[,c("title", "name")]
#> # A tibble: 15 x 2
#>    title                                            name                  
#>    <chr>                                            <chr>                 
#>  1 Monthly Mean of Water Level or Flow              hydrometric-monthly-m~
#>  2 Hydrometric Monitoring Stations                  hydrometric-stations  
#>  3 Adjusted and Homogenized Canadian Climate Data ~ ahccd-stations        
#>  4 Daily Mean of Water Level or Flow                hydrometric-daily-mean
#>  5 Adjusted and Homogenized Canadian Climate Data ~ ahccd-annual          
#>  6 Annual Maximum and Minimum Instantaneous Water ~ hydrometric-annual-pe~
#>  7 Adjusted and Homogenized Canadian Climate Data ~ ahccd-seasonal        
#>  8 Adjusted and Homogenized Canadian Climate Data ~ ahccd-trends          
#>  9 Coupled Model Intercomparison Project Phase 5 (~ cmip5                 
#> 10 MSC Monthly Data                                 climate-monthly       
#> 11 Annual Maximum and Minimum Daily Water Level or~ hydrometric-annual-st~
#> 12 1981-2010 Climate Normals                        climate-normals       
#> 13 Climate Stations                                 climate-stations      
#> 14 MSC Daily Data                                   climate-daily         
#> 15 Adjusted and Homogenized Canadian Climate Data ~ ahccd-monthly
```

For instance daily mean flow values can be queried using:

``` r
geomet_daily_mean(station_number = "10PC003")
#> No start and end dates specified. All dates available will be returned.
#> OK
#> Simple feature collection with 500 features and 12 fields
#> geometry type:  POINT
#> dimension:      XY
#> bbox:           xmin: -115.3786 ymin: 67.74028 xmax: -115.3786 ymax: 67.74028
#> epsg (SRID):    4326
#> proj4string:    +proj=longlat +datum=WGS84 +no_defs
#> # A tibble: 500 x 13
#>    id    station_number level level_symbol_en  flow flow_symbol_en
#>    <chr> <chr>          <chr> <chr>           <dbl> <chr>         
#>  1 10PC~ 10PC003        <NA>  <NA>               NA <NA>          
#>  2 10PC~ 10PC003        <NA>  <NA>               NA <NA>          
#>  3 10PC~ 10PC003        <NA>  <NA>               NA <NA>          
#>  4 10PC~ 10PC003        <NA>  <NA>               NA <NA>          
#>  5 10PC~ 10PC003        <NA>  <NA>               NA <NA>          
#>  6 10PC~ 10PC003        <NA>  <NA>               NA <NA>          
#>  7 10PC~ 10PC003        <NA>  <NA>               NA <NA>          
#>  8 10PC~ 10PC003        <NA>  <NA>               NA <NA>          
#>  9 10PC~ 10PC003        <NA>  <NA>              682 <NA>          
#> 10 10PC~ 10PC003        <NA>  <NA>              641 <NA>          
#> # ... with 490 more rows, and 7 more variables: flow_symbol_fr <chr>,
#> #   level_symbol_fr <chr>, date <date>, station_name <chr>,
#> #   identifier <chr>, prov_terr_state_loc <chr>, geometry <POINT [Â°]>
```

This returns an `sf` object by default. If you would rather simply
return a tibble you can set `as_spatial=FALSE`:

``` r
geomet_daily_mean(station_number = "10PC003", as_spatial = FALSE)
#> No start and end dates specified. All dates available will be returned.
#> OK
#> # A tibble: 1,280 x 13
#>    station_number level level_symbol_en  flow flow_symbol_en flow_symbol_fr
#>  * <chr>          <lgl> <lgl>           <dbl> <chr>          <chr>         
#>  1 10PC003        NA    NA                 NA <NA>           <NA>          
#>  2 10PC003        NA    NA                 NA <NA>           <NA>          
#>  3 10PC003        NA    NA                 NA <NA>           <NA>          
#>  4 10PC003        NA    NA                 NA <NA>           <NA>          
#>  5 10PC003        NA    NA                 NA <NA>           <NA>          
#>  6 10PC003        NA    NA                 NA <NA>           <NA>          
#>  7 10PC003        NA    NA                 NA <NA>           <NA>          
#>  8 10PC003        NA    NA                 NA <NA>           <NA>          
#>  9 10PC003        NA    NA                 NA <NA>           <NA>          
#> 10 10PC003        NA    NA                 NA <NA>           <NA>          
#> # ... with 1,270 more rows, and 7 more variables: level_symbol_fr <lgl>,
#> #   date <date>, station_name <chr>, identifier <chr>,
#> #   prov_terr_state_loc <chr>, longitude <dbl>, latitude <dbl>
```

We can also query by time ranges using the `start_date` and `end_date`
arguments:

``` r
geomet_daily_mean(station_number = "10PC003", start_date = "1983-07-11")
geomet_daily_mean(station_number = "10PC003", end_date = "1983-07-11")
```

### Getting Help or Reporting an Issue

To report bugs/issues/feature requests, please file an
[issue](https://github.com/bcgov/geometr/issues/).

### How to Contribute

If you would like to contribute to the package, please see our
[CONTRIBUTING](CONTRIBUTING.md) guidelines.

Please note that this project is released with a [Contributor Code of
Conduct](CODE_OF_CONDUCT.md). By participating in this project you agree
to abide by its terms.

### License

    Copyright 2018 Province of British Columbia
    
    Licensed under the Apache License, Version 2.0 (the &quot;License&quot;);
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at
    
    http://www.apache.org/licenses/LICENSE-2.0
    
    Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an &quot;AS IS&quot; BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and limitations under the License.

-----

*This project was created using the
[bcgovr](https://github.com/bcgov/bcgovr) package.*


NEWS.md
geometr 0.0.0.9000
=========================
### IMPROVEMENTS
* Working beta `geomet_daily_mean` function with station and date parameters
* Tests added for `geomet_daily_mean`



tests/testthat.R
library(testthat)
library(geometr)

test_check("geometr")


tests/testthat/test-geo.R
context("geomet functions will fails when no API parameters are supplied")

test_that("Test that geomet functions fails without any API parameters",{
  expect_error(geomet_daily_mean(as_spatial = TRUE), "At least one argument needs to be non NULL")
  expect_error(geomet_annual_instant_peaks(as_spatial = TRUE), "At least one argument needs to be non NULL")
  expect_error(geomet_annual_stats(as_spatial = TRUE), "At least one argument needs to be non NULL")
  expect_error(geomet_stations(as_spatial = TRUE), "At least one argument needs to be non NULL")
  expect_error(geomet_monthly_mean(as_spatial = TRUE), "At least one argument needs to be non NULL")
})


tests/testthat/test-geomet_daily_mean.R
context("geomet_daily_mean returns the correctly specified data format")

test_that("spatial is properly respected",{
  expect_is(geomet_daily_mean("02JE022", as_spatial = FALSE), "tbl_df")
  expect_is(geomet_daily_mean("02JE022"), "sf")
})


context("geomet_daily_mean respect date arguments")

test_that("two dates supplies bound the query",{
  start_date <- "1983-07-11"
  end_date <- "1983-07-21"
  d <- geomet_daily_mean("10PC003", start_date = start_date, end_date = end_date)
  expect_identical(max(d$date), as.Date(end_date))
  expect_identical(min(d$date), as.Date(start_date))
})

test_that("one end date supplies bound the query",{
    end_date <- "1983-07-21"
  d <- geomet_daily_mean("10PC003", end_date = end_date)
  expect_identical(max(d$date), as.Date(end_date))
})


test_that("one start date supplies bound the query",{
  start_date <- "1983-07-21"
  d <- geomet_daily_mean("10PC003", start_date = start_date)
  expect_identical(min(d$date), as.Date(start_date))
})

test_that("function errors when start_date is after end_date",{
  expect_error(geomet_daily_mean("08MF005", start_date = "2017-01-02", end_date = "2017-01-01"))
})


test_that("no dates supplied returns all data",{
  d <- geomet_daily_mean("10PC003")
  expect_identical(max(d$date), as.Date("1986-12-31"))
  expect_identical(min(d$date), as.Date("1983-07-01"))
  min(d$date)
})

test_that("multiple stations can be queried",{
  d <- geomet_daily_mean(c("10PC003","02JE022"))
  expect_identical(sort(unique(d$station_number)), sort(c("10PC003","02JE022")))
})

test_that("multiple stations can be queried with dates",{
  d <- geomet_daily_mean(c("10PC003","02JE022"), start_date = "1950-01-01")
  expect_identical(sort(unique(d$station_number)), sort(c("10PC003","02JE022")))
  expect_true(min(d$date) >= as.Date("1950-01-01"))
})



R/geo.R
#' Query Water survey of Canada statiosn
#'
#' Query geospatial webservice for hydrometric data
#'
#' @param STATION_NUMBER 7 digit unique station ID
#' @param as_spatial Convert to sf object
#'
#' @export

geo_stations <- function(STATION_NUMBER = NULL, as_spatial = FALSE){

  is_there_internet()

  cli <- crul::HttpClient$new(url = paste0(base_url,"/hydrometric-stations/items"))
  # do a GET request
  res <- cli$get(query = list(STATION_NUMBER = STATION_NUMBER))

  res$raise_for_status()
  # parse response to plain text (JSON in this case) - most likely you'll

  txt <- res$parse("UTF-8")

  if(as_spatial) return(sf::read_sf(txt))

  if(!as_spatial){
    # parse to json
    parsed <- jsonlite::fromJSON(txt)

    lat_long_strings <- parsed[["features"]][["geometry"]][["coordinates"]][[1]]

    cbind(parsed[["features"]][["properties"]],
          LATITUDE = lat_long_strings[2],
          LONGITUDE = lat_long_strings[1])
  }
}


geo_daily_flows <- function(STATION_NUMBER = NULL, as_spatial = FALSE){

  is_there_internet()

  (cli <- crul::HttpClient$new(url = paste0(base_url,"/hydrometric-daily-mean/items")))

  # cli_pag <- crul::Paginator$new(cli,
  #                                by = "query_params",
  #                                limit_param = "limit",
  #                                offset_param = "startindex",
  #                                limit = 10,
  #                                limit_chunk = 10)


  res <- cli$get(query = list(STATION_NUMBER = STATION_NUMBER))

  res$raise_for_status()
  # parse response to plain text (JSON in this case) - most likely you'll

  txt <- res$parse("UTF-8")

  if(as_spatial) return(sf::read_sf(txt))

  if(!as_spatial){
    # parse to json
    parsed <- jsonlite::fromJSON(txt)

    lat_long_strings <- parsed[["features"]][["geometry"]][["coordinates"]][[1]]

    cbind(parsed[["features"]][["properties"]],
          LATITUDE = lat_long_strings[2],
          LONGITUDE = lat_long_strings[1])
  }
}



R/utils.R
# Copyright 2018 Province of British Columbia
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and limitations under the License.


base_url <- function() {
  "http://geo.weather.gc.ca/geomet-beta/features/collections"
}


geomet_client <- function(geomet_param = NULL) {
  #browser()
  param_url <- sprintf("%s/%s/items", base_url(), geomet_param)

  crul::HttpClient$new(url = param_url,
                       headers = list(`User-Agent` = "https://github.com/bcgov/geometr"))
}




## Issue a query of one to get the max number of records available
num_matched_records <- function(query_list, client){
  #browser()

  query_list <- c(query_list, limit = 1)

  ## Drop any time parameter - we are just after total number for this full request
  query_list$time <- NULL

  res_max <- client$get(query = query_list)
  txt_max <- res_max$parse("UTF-8")
  jsonlite::fromJSON(txt_max)$numberMatched

}


# Argument checks ---------------------------------------------------------

check_date_format <- function(sd, ed) {

if (!is.null(sd)) {
  if (!grepl(
    "[0-9]{4}-[0-1][0-9]-[0-3][0-9]",
    sd
  )) {
    stop("Invalid date format. start_date need to be in either YYYY-MM-DD",
      call. = FALSE
    )
  }
}

if (!is.null(ed)) {
  if (!grepl(
    "[0-9]{4}-[0-1][0-9]-[0-3][0-9]",
    ed
  )) {
    stop("Invalid date format. end_date need to be in either YYYY-MM-DD",
      call. = FALSE
    )
  }
}

  if (!is.null(sd) & !is.null(ed)) {
    if (as.Date(ed) < as.Date(sd)) {
      stop("start_date is after end_date. Try swapping values.",
        call. = FALSE
      )
    }
  }

}

stop_if_all_args_null <- function() {

  ## Original idea from: https://stackoverflow.com/a/14398674/5596534

  ## Set parent environment
  parent_env <- sys.parent()

  ## Get the formals from the parent environment
  parent_formals <- formals(fun = sys.function(parent_env))

  ## Eval those formals in the parent environment
  eval_formals <- mget(names(parent_formals), sys.frame(parent_env))

  ## Subset parameters to the API can't all be null
  args_to_check_if_null <- eval_formals[names(eval_formals) %in% c("bbox","start_date","station_number")]

  ## Determine which elements are null
  ## Makes sure we don't error if other no API call params aren't called
  null_list <- vapply(args_to_check_if_null, is.null, FUN.VALUE = logical(1))

  ## Throw error is all these are null
  if(all(null_list)) stop("At least one argument needs to be non NULL", call. = FALSE)

}

multi_stations <- function(station_number){
  paste0(station_number, collapse = "|")
}


is_there_internet <- function(){
  if(!curl::has_internet()) stop("You do not appear to have a functioning internet connection")
}

handle_dates <- function(start_date, end_date){

  if(!is.null(start_date) | !is.null(end_date)){
    if(is.null(start_date)) start_date <- "1800-01-01" ## older than any other date in the db by a longshot
    if(is.null(end_date)) end_date <- Sys.Date() ## search before today's date

    return(sprintf("%s/%s", start_date, end_date))
    }

  if(is.null(start_date) && is.null(end_date)){
    message("No start and end dates specified. All dates available will be returned.")
    return(NULL)
  }
}



R/geomet.R
# Copyright 2018 Province of British Columbia
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and limitations under the License.



#' Collections available in the GeoMet webservice
#'
#' Outputs a tibble of the collections currently available via the webservice.
#'
#' @export

geomet_collection <- function() {
  cli <- crul::HttpClient$new(url = base_url())

  res <- cli$get()

  txt <- res$parse("UTF-8")

  parsed <- jsonlite::fromJSON(txt)

  parsed_tibble <- dplyr::as_tibble(parsed$collections)

  parsed_tibble$crs <- as.character(parsed_tibble$crs)

  parsed_tibble[, c("title", "name", "description", "extent", "crs", "links")]
}


R/utils-parsing.R
# Copyright 2018 Province of British Columbia
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and limitations under the License.



## Parsing helpers for a basic output
tibble_parse <- function(parsed_txt) {

  ## Iterate over paginated results
  parsed <- purrr::map_dfr(parsed_txt, parsing_helper)

  ## Extract the list column which contains the coords
  coords <- as.data.frame(do.call(rbind, parsed$coordinates))

  ## Get rid of the original list column
  parsed$coordinates <- NULL
  parsed$type <- NULL

  colnames(coords) <- c("LONGITUDE","LATITUDE")

  ## Add the new separate columns to data
  no_spatial_parsed <- cbind(parsed, coords)

  if("DATE" %in% colnames(no_spatial_parsed)){
    ## Parse date correctly
    no_spatial_parsed$DATE <- as.Date(no_spatial_parsed$DATE, format="%Y-%m-%d")
    ## Sort by date
    no_spatial_parsed <- no_spatial_parsed[order(no_spatial_parsed$DATE),]
  }

  ## All lower case
  colnames(no_spatial_parsed) <- tolower(colnames(no_spatial_parsed))

  dplyr::as_tibble(no_spatial_parsed)

}


parsing_helper <- function(x){
  list_obj <- jsonlite::fromJSON(x)

  geometry_obj <- list_obj$features$geometry

  properties_obj <- list_obj$features$properties

  cbind(properties_obj, geometry_obj)

}

spatial_parse <- function(parsed_txt, ...) {
  #browser()

  spatial_parsed <- sf::read_sf(parsed_txt, stringsAsFactors = FALSE, quiet = TRUE, ...)

  if("DATE" %in% colnames(spatial_parsed)){
    ## Sort by date
    spatial_parsed <- spatial_parsed[order(spatial_parsed$DATE),]
  }

  ## All lower case
  colnames(spatial_parsed) <- tolower(colnames(spatial_parsed))

  return(spatial_parsed)

}


R/geo-hydrometric.R
#' Query Water survey of Canada stations
#'
#' Query geospatial webservice for hydrometric data
#'
#' @param STATION_NUMBER 7 digit unique station ID
#' @param as_spatial Convert to sf object
#'
#' @export

geo_stations <- function(STATION_NUMBER = NULL, as_spatial = FALSE){

  is_there_internet()

  cli <- crul::HttpClient$new(url = paste0(base_url,"/hydrometric-stations/items"))
  # do a GET request
  res <- cli$get(query = list(STATION_NUMBER = STATION_NUMBER))

  res$raise_for_status()
  # parse response to plain text (JSON in this case) - most likely you'll

  txt <- res$parse("UTF-8")

  if(as_spatial) return(sf::read_sf(txt))

  if(!as_spatial){
    # parse to json
    parsed <- jsonlite::fromJSON(txt)

    lat_long_strings <- parsed[["features"]][["geometry"]][["coordinates"]][[1]]

    cbind(parsed[["features"]][["properties"]],
          LATITUDE = lat_long_strings[2],
          LONGITUDE = lat_long_strings[1])
  }
}


geo_daily_flows <- function(STATION_NUMBER = NULL, as_spatial = FALSE){

  is_there_internet()

  (cli <- crul::HttpClient$new(url = paste0(base_url,"/hydrometric-daily-mean/items")))

  # cli_pag <- crul::Paginator$new(cli,
  #                                by = "query_params",
  #                                limit_param = "limit",
  #                                offset_param = "startindex",
  #                                limit = 10,
  #                                limit_chunk = 10)


  res <- cli$get(query = list(STATION_NUMBER = STATION_NUMBER))

  res$raise_for_status()
  # parse response to plain text (JSON in this case) - most likely you'll

  txt <- res$parse("UTF-8")

  if(as_spatial) return(sf::read_sf(txt))

  if(!as_spatial){
    # parse to json
    parsed <- jsonlite::fromJSON(txt)

    lat_long_strings <- parsed[["features"]][["geometry"]][["coordinates"]][[1]]

    cbind(parsed[["features"]][["properties"]],
          LATITUDE = lat_long_strings[2],
          LONGITUDE = lat_long_strings[1])
  }
}



R/geomet-hydrometric.R
# Copyright 2018 Province of British Columbia
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and limitations under the License.



#' Query Water Survey of Canada data
#'
#' Query geomet webservice for hydrometric data
#'
#' @param station_number A seven digit Water Survey of Canada station number.
#' @param as_spatial Convert to sf object. Defaults to FALSE.
#'
#' @export
#'
#' @examples  geomet_stations("08MF005")

geomet_stations <- function(station_number = NULL, as_spatial = TRUE){

  stop_if_all_args_null()

  is_there_internet()

  cli <- geomet_client(geomet_param = "hydrometric-stations")

  res <- cli$get(query = list(STATION_NUMBER = station_number))

  txt <- res$parse("UTF-8")

  if(as_spatial) return(spatial_parse(txt))

  if(!as_spatial) tibble_parse(txt)

}

#' Query Water Survey of Canada data
#'
#' Query geomet webservice for hydrometric data
#'
#' @inheritParams geomet_stations
#' @param start_date Accepts either YYYY-MM-DD. If this argument is left missing, the function will return all data prior to \code{end_date}. If both are NULL all
#' available data is returned.
#' @param end_date Accepts either YYYY-MM-DD. If this argument is left missing, the function will return all data after \code{start_date}. If both are NULL all
#' available data is returned.
#' @param page_limit Default to 500. You likely will not need to change this value. This represents the maximum number of records that can be requested
#'                   for each pagination request.
#' @param verbose Defaults to FALSE. Do you want the URLs used to make the GET call to be printed?
#'
#' @export
#'
#' @examples
#' geomet_daily_mean("10PC003")
#'
#' geomet_daily_mean("10PC003", start_date = "1983-07-04")
geomet_daily_mean <- function(station_number = NULL, as_spatial = TRUE, start_date = NULL, end_date = NULL, page_limit = 500, verbose = FALSE){
  #browser()

  ## Argument Checks
  stop_if_all_args_null()
  is_there_internet()
  check_date_format(start_date, end_date)
  date_range <- handle_dates(start_date = start_date, end_date = end_date)

  ## Refine query list
  query_list <-  list(STATION_NUMBER = multi_stations(station_number),
                      time = date_range)
  ## Drop any NULLS from the list
  query_list <- Filter(Negate(is.null), query_list)


  cli <- geomet_client(geomet_param = "hydrometric-daily-mean")

  ## Total number of records available
  number_record <- num_matched_records(query_list, cli)

  (cc <- crul::Paginator$new(client = cli, by = "query_params", limit_param = "limit",
                       offset_param = "startindex", limit = number_record, limit_chunk = page_limit))

  cc$get(query = query_list)

  if(compareVersion(as.character(packageVersion("crul")), "0.6.0.9310") <= 0 && verbose) cat("Paginated URLs:\n", paste0(cc$url_fetch(),"\n"))

  txt <- cc$parse("UTF-8")


  if(as_spatial) return(spatial_parse(txt))

  if(!as_spatial) tibble_parse(txt)
}

#' Query Water Survey of Canada data
#'
#' Query geomet webservice for hydrometric data
#'
#' @inheritParams geomet_stations
#'
#' @export
#'
#' @examples  geomet_annual_stats("08MF005")
geomet_annual_stats <- function(station_number = NULL, as_spatial = FALSE) {

  stop_if_all_args_null()
  is_there_internet()

  cli <- geomet_client(geomet_param = "hydrometric-annual-statistics")

  res <- cli$get(query = list(STATION_NUMBER = station_number))

  txt <- res$parse("UTF-8")

  if(as_spatial) return(spatial_parse(txt))

  if(!as_spatial) tibble_parse(txt)
}

#' Query Water Survey of Canada data
#'
#' Query geomet webservice for hydrometric data
#'
#' @inheritParams geomet_stations
#'
#' @export
#'
#' @examples  geomet_annual_instant_peaks("08MF005")
geomet_annual_instant_peaks <- function(station_number = NULL, as_spatial = FALSE) {

  stop_if_all_args_null()
  is_there_internet()

  cli <- geomet_client(geomet_param = "hydrometric-annual-peaks")

  res <- cli$get(query = list(STATION_NUMBER = station_number))

  txt <- res$parse("UTF-8")

  if(as_spatial) return(spatial_parse(parsed_txt = txt))

  if(!as_spatial) tibble_parse(parsed_txt = txt)
}


#' Query Water Survey of Canada data
#'
#' Query geomet webservice for hydrometric data
#'
#' @inheritParams geomet_stations
#'
#' @export
#'
#' @examples geomet_monthly_mean("08MF005")
geomet_monthly_mean <- function(station_number = NULL, as_spatial = FALSE) {

  stop_if_all_args_null()
  is_there_internet()

  cli <- geomet_client(geomet_param = "hydrometric-monthly-mean")

  res <- cli$get(query = list(STATION_NUMBER = station_number))

  txt <- res$parse("UTF-8")

  if(as_spatial) return(spatial_parse(parsed_txt = txt))

  if(!as_spatial) tibble_parse(parsed_txt = txt)
}
